File: EliteFramework/EliteDevice.m
//
//  EliteCategories.m
//  BLEDrawingPad
//
//  Created by Lifelong-Study on 2015/4/21.
//  Copyright (c) 2015年 PaoYo Ding. All rights reserved.
//

#import "EliteDevice.h"
#import <sys/utsname.h>
#include "TargetConditionals.h"

// Unknown Device
#define __UNKNOWN_DEVICE          @"Unknown Device"

// iPhone
#define __IPHONE_1          @"iPhone 1"
#define __IPHONE_3G         @"iPhone 3G"
#define __IPHONE_3GS        @"iPhone 3GS"
#define __IPHONE_4          @"iPhone 4"
#define __IPHONE_4S         @"iPhone 4S"
#define __IPHONE_5          @"iPhone 5"
#define __IPHONE_5C         @"iPhone 5C"
#define __IPHONE_5S         @"iPhone 5S"
#define __IPHONE_6          @"iPhone 6"
#define __IPHONE_6_PLUS     @"iPhone 6 Plus"
#define __IPHONE_6S         @"iPhone 6S"
#define __IPHONE_6S_PLUS    @"iPhone 6S Plus"


// iPad
#define __IPAD_1            @"iPad 1"
#define __IPAD_2            @"iPad 2"
#define __IPAD_3            @"iPad 3"
#define __IPAD_4            @"iPad 4"
#define __IPAD_PRO          @"iPad Pro"
#define __IPAD_MINI_1       @"iPad mini 1"
#define __IPAD_MINI_2       @"iPad mini 2"
#define __IPAD_MINI_3       @"iPad mini 3"
#define __IPAD_MINI_4       @"iPad mini 4"
#define __IPAD_AIR_1        @"iPad Air 1"
#define __IPAD_AIR_2        @"iPad Air 2"

// iPod Touch
#define __IPOD_TOUCH_1      @"iPod Touch 1"
#define __IPOD_TOUCH_2      @"iPod Touch 2"
#define __IPOD_TOUCH_3      @"iPod Touch 3"
#define __IPOD_TOUCH_4      @"iPod Touch 4"
#define __IPOD_TOUCH_5      @"iPod Touch 5"
#define __IPOD_TOUCH_6      @"iPod Touch 6"

// Simulator
#define __IPHONE_SIMULATOR  @"iPhone Simulator"
#define __IPAD_SIMULATOR    @"iPad Simulator"


#define isStringEqual(a, b)      [(a) isEqualToString:(b)]


@implementation UIDevice (EliteDeviceCategory)

- (NSString *)getMachineIdentifier
{
    struct utsname systemInfo;
    
    uname(&systemInfo);
    
    return [NSString stringWithCString:systemInfo.machine encoding:NSUTF8StringEncoding];
}

- (EliteMachineName)machineName
{
    static EliteMachineName machineName = AppleUnknown;
    
    if (machineName) {
        return machineName;
    }
    
    NSString *machineString = self.machineString;
    
    //
    if (NO) {
    } else if ([machineString isEqualToString:__IPHONE_1]           ){ machineName = iPhone_1;
    } else if ([machineString isEqualToString:__IPHONE_3G]          ){ machineName = iPhone_3G;
    } else if ([machineString isEqualToString:__IPHONE_3GS]         ){ machineName = iPhone_3GS;
    } else if ([machineString isEqualToString:__IPHONE_4]           ){ machineName = iPhone_4;
    } else if ([machineString isEqualToString:__IPHONE_4S]          ){ machineName = iPhone_4S;
    } else if ([machineString isEqualToString:__IPHONE_5]           ){ machineName = iPhone_5;
    } else if ([machineString isEqualToString:__IPHONE_5C]          ){ machineName = iPhone_5C;
    } else if ([machineString isEqualToString:__IPHONE_5S]          ){ machineName = iPhone_5S;
    } else if ([machineString isEqualToString:__IPHONE_6]           ){ machineName = iPhone_6;
    } else if ([machineString isEqualToString:__IPHONE_6_PLUS]      ){ machineName = iPhone_6_Plus;
    } else if ([machineString isEqualToString:__IPHONE_6S]          ){ machineName = iPhone_6S;
    } else if ([machineString isEqualToString:__IPHONE_6S_PLUS]     ){ machineName = iPhone_6S_Plus;
        
    //
    } else if ([machineString isEqualToString:__IPAD_1]             ){ machineName = iPad_1;
    } else if ([machineString isEqualToString:__IPAD_2]             ){ machineName = iPad_2;
    } else if ([machineString isEqualToString:__IPAD_3]             ){ machineName = iPad_3;
    } else if ([machineString isEqualToString:__IPAD_4]             ){ machineName = iPad_4;
    } else if ([machineString isEqualToString:__IPAD_MINI_1]        ){ machineName = iPad_mini_1;
    } else if ([machineString isEqualToString:__IPAD_MINI_2]        ){ machineName = iPad_mini_2;
    } else if ([machineString isEqualToString:__IPAD_MINI_3]        ){ machineName = iPad_mini_3;
    } else if ([machineString isEqualToString:__IPAD_MINI_4]        ){ machineName = iPad_mini_4;
    } else if ([machineString isEqualToString:__IPAD_AIR_1]         ){ machineName = iPad_Air_1;
    } else if ([machineString isEqualToString:__IPAD_AIR_2]         ){ machineName = iPad_Air_2;
    } else if ([machineString isEqualToString:__IPAD_PRO]           ){ machineName = iPad_Pro;
    
 
    // iPod Touch
    } else if ([machineString isEqualToString:__IPOD_TOUCH_1]      ){ machineName = iPod_Touch_1;
    } else if ([machineString isEqualToString:__IPOD_TOUCH_2]      ){ machineName = iPod_Touch_2;
    } else if ([machineString isEqualToString:__IPOD_TOUCH_3]      ){ machineName = iPod_Touch_3;
    } else if ([machineString isEqualToString:__IPOD_TOUCH_4]      ){ machineName = iPod_Touch_4;
    } else if ([machineString isEqualToString:__IPOD_TOUCH_5]      ){ machineName = iPod_Touch_5;
    
    // Simulator
    } else if ([machineString isEqualToString:__IPHONE_SIMULATOR]   ){ machineName = iOSSimulator;
    } else if ([machineString isEqualToString:__IPAD_SIMULATOR]     ){ machineName = iOSSimulator;
    
    //
    } else    { NSLog(@"無法辨識的 machineString: %@", machineString); }
    
    return machineName;
}

- (EliteMachineType)type
{
//    _type = type;
    
    static EliteMachineType _type = MachineTypeUnknown;

    if (_type != MachineTypeUnknown) {
        return _type;
    }
    
//    if (NO) {
//    } else if ([self.machineString containsString:@"i386"]   ){     _type = MachineTypeWatch;
//    } else if ([self.machineString containsString:@"iPhone"] ){     _type = MachineTypeiPhone;
//    } else if ([self.machineString containsString:@"iPad"]   ){     _type = MachineTypeiPad;
//    } else if ([self.machineString containsString:@"iPod"]   ){     _type = MachineTypeiPod;
//    } else if ([self.machineString containsString:@"AppleTV"]){     _type = MachineTypeTV;
//    } else if ([self.machineString containsString:@"Watch"]  ){     _type = MachineTypeWatch;
//    } else {                                                        _type = MachineTypeUnknown;
//    }
    
    if (NO) {
    } else if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPhone) {    _type = MachineTypeiPhone;
    } else if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {      _type = MachineTypeiPad;
    } else if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomTV) {      _type = MachineTypeTV;
//    } else if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomCarPlay) {      _type = iPad;
//    } else if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {      _type = iPad;
    }
    
    
    return _type;
}

- (NSString *)machineString
{
    static NSString *string = __UNKNOWN_DEVICE;;
    
    if (string) {
        return string;
    }
    
    NSString *identifier = [self getMachineIdentifier];
    
    NSLog(@"Device identifier = %@", identifier);
    // Model information retrieved from http://theiphonewiki.com/wiki/Models
    
    // iPhone
    if (NO) {
    } else if (isStringEqual(identifier, @"iPhone1,1")){ string = __IPHONE_1;         // iPhone 2G GSM
    } else if (isStringEqual(identifier, @"iPhone1,2")){ string = __IPHONE_3G;        // iPhone 3G GSM
    } else if (isStringEqual(identifier, @"iPhone2,1")){ string = __IPHONE_3GS;       // iPhone 3GS GSM
    } else if (isStringEqual(identifier, @"iPhone3,1")){ string = __IPHONE_4;         // iPhone 4 GSM
    } else if (isStringEqual(identifier, @"iPhone3,2")){ string = __IPHONE_4;         // iPhone 4 GSM Rev A
    } else if (isStringEqual(identifier, @"iPhone3,3")){ string = __IPHONE_4;         // iPhone 4 CDMA
    } else if (isStringEqual(identifier, @"iPhone4,1")){ string = __IPHONE_4S;        // iPhone 4S
    } else if (isStringEqual(identifier, @"iPhone5,1")){ string = __IPHONE_5;         // iPhone 5 GSM
    } else if (isStringEqual(identifier, @"iPhone5,2")){ string = __IPHONE_5;         // iPhone 5 GSM+CDMA
    } else if (isStringEqual(identifier, @"iPhone5,3")){ string = __IPHONE_5C;        // iPhone 5C GSM
    } else if (isStringEqual(identifier, @"iPhone5,4")){ string = __IPHONE_5C;        // iPhone 5C Global
    } else if (isStringEqual(identifier, @"iPhone6,1")){ string = __IPHONE_5S;        // iPhone 5S GSM
    } else if (isStringEqual(identifier, @"iPhone6,2")){ string = __IPHONE_5S;        // iPhone 5S Global
    } else if (isStringEqual(identifier, @"iPhone7,1")){ string = __IPHONE_6_PLUS;    //
    } else if (isStringEqual(identifier, @"iPhone7,2")){ string = __IPHONE_6;         //
    } else if (isStringEqual(identifier, @"iPhone8,1")){ string = __IPHONE_6S;        //
    } else if (isStringEqual(identifier, @"iPhone8,2")){ string = __IPHONE_6S_PLUS;   //
        
    // iPad
    } else if (isStringEqual(identifier, @"iPad2,1")  ){ string = __IPAD_1;           // iPad 1 Wi-Fi/GSM
    } else if (isStringEqual(identifier, @"iPad2,1")  ){ string = __IPAD_2;           // iPad 2 Wi-Fi
    } else if (isStringEqual(identifier, @"iPad2,2")  ){ string = __IPAD_2;           // iPad 2 GSM
    } else if (isStringEqual(identifier, @"iPad2,3")  ){ string = __IPAD_2;           // iPad 2 CDMA
    } else if (isStringEqual(identifier, @"iPad2,4")  ){ string = __IPAD_2;           // iPad 2 Wi-Fi Rev A
    } else if (isStringEqual(identifier, @"iPad2,5")  ){ string = __IPAD_MINI_1;      // iPad Mini 1G Wi-Fi
    } else if (isStringEqual(identifier, @"iPad2,6")  ){ string = __IPAD_MINI_1;      // iPad Mini 1G GSM
    } else if (isStringEqual(identifier, @"iPad2,7")  ){ string = __IPAD_MINI_1;      // iPad Mini 1G GSM+CDMA
    } else if (isStringEqual(identifier, @"iPad3,1")  ){ string = __IPAD_3;           // iPad 3 Wi-Fi
    } else if (isStringEqual(identifier, @"iPad3,2")  ){ string = __IPAD_3;           // iPad 3 GSM+CDMA
    } else if (isStringEqual(identifier, @"iPad3,3")  ){ string = __IPAD_3;           // iPad 3 GSM
    } else if (isStringEqual(identifier, @"iPad3,4")  ){ string = __IPAD_4;           // iPad 4 Wi-Fi
    } else if (isStringEqual(identifier, @"iPad3,5")  ){ string = __IPAD_4;           // iPad 4 GSM
    } else if (isStringEqual(identifier, @"iPad3,6")  ){ string = __IPAD_4;           // iPad 4 GSM+CDMA
    } else if (isStringEqual(identifier, @"iPad4,1")  ){ string = __IPAD_AIR_1;       // iPad Air Wi-Fi
    } else if (isStringEqual(identifier, @"iPad4,2")  ){ string = __IPAD_AIR_1;       // iPad Air Cellular
    } else if (isStringEqual(identifier, @"iPad4,3")  ){ string = __IPAD_AIR_1;       // iPad Air Cellular
    } else if (isStringEqual(identifier, @"iPad4,4")  ){ string = __IPAD_MINI_2;      // iPad Mini 2
    } else if (isStringEqual(identifier, @"iPad4,5")  ){ string = __IPAD_MINI_2;      // iPad Mini 2
    } else if (isStringEqual(identifier, @"iPad4,6")  ){ string = __IPAD_MINI_2;      // iPad Mini 2
    } else if (isStringEqual(identifier, @"iPad4,7")  ){ string = __IPAD_MINI_3;      // iPad Mini 3
    } else if (isStringEqual(identifier, @"iPad4,8")  ){ string = __IPAD_MINI_3;      // iPad Mini 3
    } else if (isStringEqual(identifier, @"iPad4,9")  ){ string = __IPAD_MINI_3;      // iPad Mini 3
    } else if (isStringEqual(identifier, @"iPad5,1")  ){ string = __IPAD_MINI_4;      // iPad Mini 4
    } else if (isStringEqual(identifier, @"iPad5,2")  ){ string = __IPAD_MINI_4;      // iPad Mini 4
    } else if (isStringEqual(identifier, @"iPad5,3")  ){ string = __IPAD_AIR_2;       // iPad Mini 4
    } else if (isStringEqual(identifier, @"iPad5,4")  ){ string = __IPAD_AIR_2;       // iPad Mini 4
    } else if (isStringEqual(identifier, @"iPad6,7")  ){ string = __IPAD_PRO;         //
    } else if (isStringEqual(identifier, @"iPad6,8")  ){ string = __IPAD_PRO;         //
        
    // iPod Touch
    } else if (isStringEqual(identifier, @"iPod1,1")  ){ string = __IPOD_TOUCH_1;    // iPod Touch 1
    } else if (isStringEqual(identifier, @"iPod2,1")  ){ string = __IPOD_TOUCH_2;    // iPod Touch 2
    } else if (isStringEqual(identifier, @"iPod3,1")  ){ string = __IPOD_TOUCH_3;    // iPod Touch 3
    } else if (isStringEqual(identifier, @"iPod4,1")  ){ string = __IPOD_TOUCH_4;    // iPod Touch 4
    } else if (isStringEqual(identifier, @"iPod5,1")  ){ string = __IPOD_TOUCH_5;    // iPod Touch 5
    } else if (isStringEqual(identifier, @"iPod6,1")  ){ string = __IPOD_TOUCH_6;    // iPod Touch 6
        
    // Simulator
    } else if (isStringEqual(identifier, @"i386")     ){
        if (NO) {
        } else if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPhone) {    string = __IPHONE_SIMULATOR;
        } else if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad) {      string = __IPAD_SIMULATOR;
        }
    }
    
    return string;
}

@end

File: EliteFramework/EliteDevice.swift
//
//  EliteDevice.swift
//  EliteFramework
//
//  Created by Lifelong-Study on 2016/2/14.
//  Copyright © 2016年 Lifelong-Study. All rights reserved.
//

import UIKit


public enum MachineName {
    // Unknown Device
    case unknown_Device
    
    // iPhone
    case iPhone_1
    case iPhone_3G
    case iPhone_3GS
    case iPhone_4
    case iPhone_4S
    case iPhone_5
    case iPhone_5C
    case iPhone_5S
    case iPhone_6
    case iPhone_6_Plus
    case iPhone_6S
    case iPhone_6S_Plus
    case iPhone_SE
    
    
    // iPad
    case iPad_1
    case iPad_2
    case iPad_3
    case iPad_4
    case iPad_Pro
    case iPad_mini_1
    case iPad_mini_2
    case iPad_mini_3
    case iPad_mini_4
    case iPad_Air_1
    case iPad_Air_2
    
    // iPod Touch
    case iPod_Touch_1
    case iPod_Touch_2
    case iPod_Touch_3
    case iPod_Touch_4
    case iPod_Touch_5
    case iPod_Touch_6
    
    // Simulator
    case iPhone_Simulator
    case iPad_Simulator
}


public extension UIDevice {
//    let swf: MachineName!
    
    public class func machineName() -> MachineName {
        let screenSize = UIScreen.main.bounds.size
        
        
        if false {
        } else if screenSize.width == 320 && screenSize.height == 480 {
            return .iPhone_4
        } else if screenSize.width == 320 && screenSize.height == 568 {
            return .iPhone_5
        } else if screenSize.width == 375 && screenSize.height == 667 {
            return .iPhone_6
        } else if screenSize.width == 414 && screenSize.height == 736 {
            return .iPhone_6_Plus
        } else {
            return .iPhone_6_Plus
        }
    }
    
    fileprivate static var _model: MachineName? = nil
    
    public class var model: MachineName? {
    
        if UIDevice._model != nil {
            return UIDevice._model
        }

        switch UIDevice.getMachineIdentifier() {
            case "iPhone1,1":   UIDevice._model = .iPhone_1         // iPhone 2G GSM
            case "iPhone1,2":   UIDevice._model = .iPhone_3G        // iPhone 3G GSM
            case "iPhone2,1":   UIDevice._model = .iPhone_3GS       // iPhone 3GS GSM
            case "iPhone3,1":   UIDevice._model = .iPhone_4         // iPhone 4 GSM
            case "iPhone3,2":   UIDevice._model = .iPhone_4         // iPhone 4 GSM Rev A
            case "iPhone3,3":   UIDevice._model = .iPhone_4         // iPhone 4 CDMA
            case "iPhone4,1":   UIDevice._model = .iPhone_4S        // iPhone 4S
            case "iPhone5,1":   UIDevice._model = .iPhone_5         // iPhone 5 GSM
            case "iPhone5,2":   UIDevice._model = .iPhone_5         // iPhone 5 GSM+CDMA
            case "iPhone5,3":   UIDevice._model = .iPhone_5C        // iPhone 5C GSM
            case "iPhone5,4":   UIDevice._model = .iPhone_5C        // iPhone 5C Global
            case "iPhone6,1":   UIDevice._model = .iPhone_5S        // iPhone 5S GSM
            case "iPhone6,2":   UIDevice._model = .iPhone_5S        // iPhone 5S Global
            case "iPhone7,1":   UIDevice._model = .iPhone_6_Plus    //
            case "iPhone7,2":   UIDevice._model = .iPhone_6         //
            case "iPhone8,1":   UIDevice._model = .iPhone_6S        //
            case "iPhone8,2":   UIDevice._model = .iPhone_6S_Plus   //
            case "iPhone8,4":   UIDevice._model = .iPhone_SE        //
            
            // iPad
            case "iPad2,1":     UIDevice._model = .iPad_1           // iPad 1 Wi-Fi/GSM
            case "iPad2,1":     UIDevice._model = .iPad_2           // iPad 2 Wi-Fi
            case "iPad2,2":     UIDevice._model = .iPad_2           // iPad 2 GSM
            case "iPad2,3":     UIDevice._model = .iPad_2           // iPad 2 CDMA
            case "iPad2,4":     UIDevice._model = .iPad_2           // iPad 2 Wi-Fi Rev A
            case "iPad2,5":     UIDevice._model = .iPad_mini_1      // iPad Mini 1G Wi-Fi
            case "iPad2,6":     UIDevice._model = .iPad_mini_1      // iPad Mini 1G GSM
            case "iPad2,7":     UIDevice._model = .iPad_mini_1      // iPad Mini 1G GSM+CDMA
            case "iPad3,1":     UIDevice._model = .iPad_3           // iPad 3 Wi-Fi
            case "iPad3,2":     UIDevice._model = .iPad_3           // iPad 3 GSM+CDMA
            case "iPad3,3":     UIDevice._model = .iPad_3           // iPad 3 GSM
            case "iPad3,4":     UIDevice._model = .iPad_4           // iPad 4 Wi-Fi
            case "iPad3,5":     UIDevice._model = .iPad_4           // iPad 4 GSM
            case "iPad3,6":     UIDevice._model = .iPad_4           // iPad 4 GSM+CDMA
            case "iPad4,1":     UIDevice._model = .iPad_Air_1       // iPad Air Wi-Fi
            case "iPad4,2":     UIDevice._model = .iPad_Air_1       // iPad Air Cellular
            case "iPad4,3":     UIDevice._model = .iPad_Air_1       // iPad Air Cellular
            case "iPad4,4":     UIDevice._model = .iPad_mini_2      // iPad Mini 2
            case "iPad4,5":     UIDevice._model = .iPad_mini_2      // iPad Mini 2
            case "iPad4,6":     UIDevice._model = .iPad_mini_2      // iPad Mini 2
            case "iPad4,7":     UIDevice._model = .iPad_mini_3      // iPad Mini 3
            case "iPad4,8":     UIDevice._model = .iPad_mini_3      // iPad Mini 3
            case "iPad4,9":     UIDevice._model = .iPad_mini_3      // iPad Mini 3
            case "iPad5,1":     UIDevice._model = .iPad_mini_4      // iPad Mini 4
            case "iPad5,2":     UIDevice._model = .iPad_mini_4      // iPad Mini 4
            case "iPad5,3":     UIDevice._model = .iPad_Air_2       // iPad Air 2
            case "iPad5,4":     UIDevice._model = .iPad_Air_2       // iPad Air 2
            case "iPad6,7":     UIDevice._model = .iPad_Pro         //
            case "iPad6,8":     UIDevice._model = .iPad_Pro         //
            
            // iPod Touch
            case "iPod1,1":     UIDevice._model = .iPod_Touch_1    // iPod Touch 1
            case "iPod2,1":     UIDevice._model = .iPod_Touch_2    // iPod Touch 2
            case "iPod3,1":     UIDevice._model = .iPod_Touch_3    // iPod Touch 3
            case "iPod4,1":     UIDevice._model = .iPod_Touch_4    // iPod Touch 4
            case "iPod5,1":     UIDevice._model = .iPod_Touch_5    // iPod Touch 5
            case "iPod6,1":     UIDevice._model = .iPod_Touch_6    // iPod Touch 6
            
            // Simulator
            case "i386":        if false {
                                } else if UI_USER_INTERFACE_IDIOM() == .phone {     UIDevice._model = .iPhone_Simulator
                                } else if UI_USER_INTERFACE_IDIOM() == .pad   {     UIDevice._model = .iPad_Simulator
                                }
            
            //
            default:            UIDevice._model = .unknown_Device
        }
        
        return UIDevice._model
    }
    
    class func getMachineIdentifier() -> String {
        
        var systemInfo = utsname()
        
        uname(&systemInfo)
        
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        
        let identifier = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8 , value != 0 else { return identifier }
            
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
        
        return identifier
    }
}
